pallets/automation-price/src/lib.rs

```diff
@@ -0,0 +1,699 @@
+				if asset_sudo != who {
+					Err(Error::<T>::InvalidAssetSudo)?
+				}
+			}
+			let fee = <BalanceOf<T>>::saturated_from(1_000_000_000_000u64);
+			T::FeeHandler::can_pay_fee(&who.clone(), fee.clone())
+				.map_err(|_| Error::<T>::InsufficientBalance)?;
+			if let Some(asset_target_price) = Self::get_asset_target_price(asset.clone()) {
+				let last_asset_price: u128 = match Self::get_asset_price(asset.clone()) {
+					None => Err(Error::<T>::AssetNotSupported)?,
+					Some(asset_price) => asset_price,
+				};
+				let asset_update_percentage = Self::calculate_asset_percentage(value, asset_target_price)
+					.saturating_add(1);
+				let asset_last_percentage = 0;
+				if value > last_asset_price {
+					Self::move_scheduled_tasks(asset.clone(), asset_last_percentage, asset_update_percentage, 1)?;
+				} else {
+					Self::move_scheduled_tasks(asset.clone(), asset_last_percentage, asset_update_percentage, 0)?;
+				}
+				AssetPrices::<T>::insert(asset.clone(), value);
+				T::FeeHandler::withdraw_fee(&who, fee.clone())
+					.map_err(|_| Error::<T>::LiquidityRestrictions)?;
+				Self::deposit_event(Event::AssetUpdated { asset });
+			} else {
+				Err(Error::<T>::AssetNotSupported)?
+			}
+			Ok(().into())
+		}
+
+		/// Delete an asset
+		///
+		/// # Parameters
+		/// * `asset`: asset type
+		/// * `directions`: number of directions of data input. (up, down, ?)
+		///
+		/// # Errors
+		#[pallet::weight(<T as Config>::WeightInfo::delete_asset_extrinsic())]
+		#[transactional]
+		pub fn delete_asset(
+			origin: OriginFor<T>,
+			asset: Vec<u8>,
+		) -> DispatchResult {
+			// TODO: needs fees if opened up to non-sudo
+			ensure_root(origin)?;
+			if let Some(_asset_target_price) = Self::get_asset_target_price(asset.clone()) {
+				AssetTargetPrices::<T>::remove(asset.clone());
+				AssetPrices::<T>::remove(asset.clone());
+				AssetMetadata::<T>::remove(asset.clone());
+				Self::delete_asset_tasks(asset.clone());
+				Self::deposit_event(Event::AssetDeleted { asset });
+			} else {
+				Err(Error::<T>::AssetNotSupported)?
+			}
+			Ok(().into())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Trigger tasks for the block time.
+		///
+		/// Complete as many tasks as possible given the maximum weight.
+		pub fn trigger_tasks(max_weight: Weight) -> Weight {
+			let mut weight_left: Weight = max_weight;
+			let check_time_and_deletion_weight = T::DbWeight::get().reads(2 as Weight);
+			if weight_left < check_time_and_deletion_weight {
+				return weight_left
+			}
+
+			// remove assets as necessary
+			let current_time_slot = match Self::get_current_time_slot() {
+				Ok(time_slot) => time_slot,
+				Err(_) => return weight_left,
+			};
+			if let Some(scheduled_deletion_assets) = Self::get_scheduled_asset_period_reset(current_time_slot) {
+				// delete assets' tasks
+				let asset_reset_weight = <T as Config>::WeightInfo::reset_asset(scheduled_deletion_assets.len().try_into().unwrap());
```

Comment:
no weight calc for delete_asset_tasks or update_asset_reset because it's wrapped up in here

Reply:
also, is this a safe unwrap? i think so? scheduled_deletion_assets is usize where as reset_asset is expecting a u64? so it should be fine to unwrap, right? i doubt there will be more than 2^64 assets in our DB 😅